prompt.go

```diff
@@ -346,18 +383,108 @@ func (p *Prompt) tearDown() {
 	p.renderer.TearDown()
 }
 
-func (p *Prompt) turnOffReverseSearch(eraseLinesCnt int) {
-	for i := 0; i < eraseLinesCnt; i++ {
-		p.renderer.out.CursorUp(1)
-		p.renderer.out.EraseLine()
+// getCmdToRender builds command to render.
+// returns (buffer with command to render, prefix length in bytes)
+func (prompt *Prompt) getCmdToRender() (cmd *Buffer, prefixLen int) {
+	input := prompt.buf.Text()
+	prefix := prompt.getCurrentPrefix()
+	cmdBuf := NewBuffer()
+	cmdBuf.InsertText(prefix, false, true)
+	if prompt.inReverseSearchMode() {
+		cmdBuf.InsertText(prompt.reverseSearch.matchedCmd, false, true)
+	} else {
+		cmdBuf.InsertText(input, false, true)
+		cmdBuf.setCursorPosition(len(prefix) + prompt.buf.cursorPosition)
 	}
+	return cmdBuf, len(prefix)
+}
 
-	p.renderer.out.CursorBackward(maxBackwardColumn)
+// getCurrentPrefix returns current prefix.
+// If reverse search is enabled, its prefix extracted.
+// If live-prefix is enabled, return live-prefix.
+func (prompt *Prompt) getCurrentPrefix() string {
+	if prompt.inReverseSearchMode() {
+		rsPrefixFmt := matchSearchPrefixFmt
+		if prompt.reverseSearch.matchedIndex == -1 {
+			rsPrefixFmt = failSearchPrefixFmt
+		}
+		return fmt.Sprintf(rsPrefixFmt, prompt.buf.Text())
+	}
+	if prefix, ok := prompt.livePrefixCallback(); ok {
+		return prefix
+	}
+	return prompt.prefix
+}
+
+// onInputUpdate does necessary actions at the input update moment.
+func (prompt *Prompt) onInputUpdate() {
+	prompt.buf = prompt.buf.ReplaceTabs(defaultTabWidth)
+	if prompt.inReverseSearchMode() {
+		prompt.reverseSearch.update(prompt.buf.Text())
+		return
+	}
+	prompt.history.SetCurrentCmd(prompt.buf.Text())
+	prompt.completion.Update(*prompt.buf.Document())
+}
+
+// render renders current prompt state to the attached renderer,
+// updates current cursor position.
+func (prompt *Prompt) render(event int) {
+	ctx := prompt.fillCtx(event)
+	prompt.cursor, prompt.endCursor = prompt.renderer.Render(ctx)
+}
+
+// inReverseSearchMode returns true if the prompt is in reverse-search mode.
+func (p *Prompt) inReverseSearchMode() bool {
+	return p.reverseSearch != nil
+}
+
+// enableReverseSearch enables reverse-search mode.
+func (p *Prompt) enableReverseSearch() {
+	if !p.reverseSearchOpt {
+		return
+	}
+
+	p.buf = NewBuffer()
+	p.reverseSearch = NewReverseSearch(p.history)
+}
+
+// disableReverseSearch disables reverse-search mode,
+// sets history pointer to the last matched command.
+func (p *Prompt) disableReverseSearch() {
+	if !p.reverseSearchOpt {
+		return
+	}
 
+	matchedIndex := p.reverseSearch.matchedIndex
```

Comment:
It was assumed that it was called when reverse-search was turned off. But, it is more logical to check, because we are already checking another flag inside this function. So the check is added.